jetty-util/src/main/java/org/eclipse/jetty/util/ssl/SslKeyStoreScanner.java

```diff
@@ -0,0 +1,145 @@
+//
+//  ========================================================================
+//  Copyright (c) 1995-2020 Mort Bay Consulting Pty Ltd and others.
+//  ------------------------------------------------------------------------
+//  All rights reserved. This program and the accompanying materials
+//  are made available under the terms of the Eclipse Public License v1.0
+//  and Apache License v2.0 which accompanies this distribution.
+//
+//      The Eclipse Public License is available at
+//      http://www.eclipse.org/legal/epl-v10.html
+//
+//      The Apache License v2.0 is available at
+//      http://www.opensource.org/licenses/apache2.0.php
+//
+//  You may elect to redistribute this code under either of these licenses.
+//  ========================================================================
+//
+
+package org.eclipse.jetty.util.ssl;
+
+import java.io.File;
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.List;
+
+import org.eclipse.jetty.util.Scanner;
+import org.eclipse.jetty.util.annotation.ManagedAttribute;
+import org.eclipse.jetty.util.annotation.ManagedOperation;
+import org.eclipse.jetty.util.component.AbstractLifeCycle;
+import org.eclipse.jetty.util.log.Log;
+import org.eclipse.jetty.util.log.Logger;
+
+public class SslKeyStoreScanner extends AbstractLifeCycle implements Scanner.DiscreteListener
+{
+    private static final Logger LOG = Log.getLogger(SslKeyStoreScanner.class);
+
+    private final SslContextFactory sslContextFactory;
+    private final File keystoreFile;
+    private final List<File> files = new ArrayList<>();
+    private Scanner _scanner;
+    private int _scanInterval = 1;
+
+    public SslKeyStoreScanner(SslContextFactory sslContextFactory)
+    {
+        this.sslContextFactory = sslContextFactory;
+        this.keystoreFile = new File(URI.create(sslContextFactory.getKeyStorePath())); // getKeyStorePath is giving url instead of path?
```

Comment:
Perhaps try ... `sslContextFactory.getKeyStoreResource().getFile()`?

Reply:
How do you handle it when there is no KeyStore Resource/Path/File?

Such as when the user uses `SslContextFactory.setKeyStore(KeyStore)` or `SslContextFactory.setKeyStoreResource(Resource)` on a resource that is not backed by a File?

Reply:
If there is no keystore file we will hit one of these `IllegalArgumentException`s and have an error during startup. I think this is fine as its bad configuration to have this when the keystore is not backed by a file.